Mathlib/Computability/QueryComplexity/Defs.lean

```diff
@@ -0,0 +1,116 @@
+/-
+Copyright (c) 2025 Tomaz Mascarenhas. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Geoffrey Irving, Tomaz Mascarenhas, Eric Wieser
+-/
+import Mathlib.Algebra.Group.Pi.Basic
+import Mathlib.Data.Set.Insert
+
+/-!
+# Definition of a model of computation based on oracles.
+
+This module defines an abstraction of computation with oracles, enabling the analysis of upper
+bounds on the query complexity of algorithms. It also provides a Monad instance for it.
+
+Note that this is the deterministic version. In the future, the stochastic version will
+also be ported, which enables the possibility of proving lower bounds on the query complexity
+of algorithms. In order to facilitate this future integration, the design of the types
+in this module is based on the original stochastic version. Source:
+  https://github.com/girving/debate/blob/862fdb1cf55df0d541b802bdb1e672d724df6398/Comp/Oracle.lean
+
+## Main Definitions
+
+* Comp (ι : Type) {I : Type} (s : Set I) (α : Type) : Type
+* Comp.run : Comp ι s α → (I → Oracle ι) → α × (I → ℕ)
+-/
+
+open Set
+
+universe u
+
+variable {I : Type*}
+variable {ι : I → Type*}
+variable {s t : Set I}
+variable {α β γ : Type*}
+variable {ω : {i : I} → ι i → Type*}
+
+namespace QueryComplexity
+
+/-- A deterministic oracle is a dependent map -/
+def Oracle (α : Type*) (β : α → Type*) : Type _ := (x : α) → (β x)
+
+/-- An `Oracle` that always returns `Bool` -/
+abbrev BOracle (α : Type*) : Type _ := Oracle α fun _ ↦ Bool
+
+/-- A deterministic computation that make decisions by querying oracles. A computation is either a
+pure value or the identifier of an oracle (`o`) drawn from a predefined set (`s`), a value to
+be queried by the oracle (`i`) and a dependent selection function that determines which oracle to
+run, depending on the result of the query. -/
```

Comment:
This isn't true; it doesn't determine which _oracle_ to run, but rather which _continuation_ `Comp` to run.